### PR TITLE
support stop-loss/take-profit limit orders

### PR DESCRIPTION
**Pull request by vinta**
Require https://github.com/perpetual-protocol/perp-oracle-contract/pull/33

### PR Reminders
---

1. **Contract**: make sure the code follows our convention, ex: naming, explicit returns, etc.; if uncertain, discuss with others

2. **Test**: make sure tests can cover most normal and edge cases; if not, open follow-up tickets. Also, look out for failed tests and fix them!

3. **CHANGELOG.md**: update when external interfaces are changed

4. Workflow: 
    - Github: assign the pr to yourself; if pairing, can merge directly; else, assign someone to help review
    - Asana: assign the corresponding ticket to yourself and leave the pr link on it for easier follow-ups
    - Discord: if someone is mentioned in this pr, tag on Discord

### CODE REVIEW

File: contracts/limitOrder/LimitOrderBook.sol [hunk 1]
```diff
@@ -219,4 +228,66 @@ contract LimitOrderBook is
...
+        // https://docs.chain.link/docs/historical-price-data/
+
+        // LOB_IRI: Invalid Round Id
+        require(order.roundIdWhenCreated > 0 && roundIdWhenTriggered > order.roundIdWhenCreated, "LOB_IRI");
+
+        // LOB_ITP: Invalid Trigger Price
+        require(order.triggerPrice > 0, "LOB_ITP");
+
+        // NOTE: we can only support stop limit order for markets that use ChainlinkPriceFeed
+        uint256 triggeredPrice = _getPriceByRoundId(order.baseToken, roundIdWhenTriggered);
```
Comment (vinta) [inline contracts/limitOrder/LimitOrderBook.sol:264]: What will happen when calling `_getPriceByRoundId()` and the market is using BandPriceFeed?

Comment (tienshaoku) [reply contracts/limitOrder/LimitOrderBook.sol:264]: Is this a TODO?
If yes, I guess we'll have to finish this TODO before merging this pr!

I guess an (easy?) approach can be to keep a list of ChainlinkPriceFeed and verify the priceFeed using is on the list.

File: contracts/limitOrder/LimitOrderBook.sol [hunk 2]
```diff
@@ -112,13 +113,21 @@ contract LimitOrderBook is
...
+
+            // if trader has no position, order will get reverted
+            // if trader has short position, trader can only open a long position
             // => oldTakerPositionSize < 0 != order.isBaseToQuote => true != false
-            // if trader has long position, he/she can only open a short position
+            // if trader has long position, trader can only open a short position
             // => oldTakerPositionSize < 0 != order.isBaseToQuote => false != true
         }
 
+        if (
```
Comment (tienshaoku) [inline contracts/limitOrder/LimitOrderBook.sol:124]: Perhaps we can merge as `order.orderType != ILimitOrderBook.OrderType.LimitOrder`; or is there any concern?

File: contracts/limitOrder/LimitOrderBook.sol [hunk 3]
```diff
@@ -219,4 +228,66 @@ contract LimitOrderBook is
...
+        return price.mul(10**(toDecimals.sub(fromDecimals)));
+    }
+
+    function _getPriceByRoundId(address baseToken, uint80 roundId) internal view returns (uint256) {
+        ChainlinkPriceFeed chainlinkPriceFeed = ChainlinkPriceFeed(IBaseToken(baseToken).getPriceFeed());
+        (uint256 price, ) = chainlinkPriceFeed.getRoundData(roundId);
+        return _formatDecimals(price, chainlinkPriceFeed.decimals(), 18);
+    }
+
+    function _verifyTriggerPrice(LimitOrder memory order, uint80 roundIdWhenTriggered) internal view {
```
Comment (tienshaoku) [inline contracts/limitOrder/LimitOrderBook.sol:253]: Wonder why this is internal but `verifySigner` is public? Are we sure that no one needs to use this on their own bots?

File: contracts/interface/ILimitOrderBook.sol [hunk 4]
```diff
@@ -99,8 +98,8 @@ interface ILimitOrderBook {
 
     /// @param order LimitOrder struct
     /// @param signature The EIP-712 signature of `order` generated from `eth_signTypedData_V4`
-    /// @param roundIdWhenTriggered The oracle `roundId` when triggerPrice is satisfied
-    // Only avaliable if orderType is StopLimitOrder, otherwise set to 0
+    /// @param roundIdWhenTriggered Chainlink `roundId` when triggerPrice is satisfied
+    // Only avaliable if orderType is StopLimitOrder/TakeProfitLimitOrder, otherwise set to 0
```
Comment (tienshaoku) [inline contracts/interface/ILimitOrderBook.sol:102]: Minor typo: `available`

File: contracts/limitOrder/LimitOrderBook.sol [hunk 5]
```diff
@@ -30,6 +33,8 @@ contract LimitOrderBook is
     using PerpMath for uint256;
     using PerpSafeCast for uint256;
     using SignedSafeMathUpgradeable for int256;
+    using SafeMathUpgradeable for uint256;
+    using SafeMathUpgradeable for uint8;
 
     // NOTE: cannot use `OrderType orderType` here, use `uint8 orderType` instead
     // solhint-disable-next-line func-name-mixedcase
```
Comment (tienshaoku) [inline contracts/limitOrder/LimitOrderBook.sol:40]: I'm not really familiar with pre-computing hash but I thought we could compute this locally and paste the hash directly and can thus save this long line

File: contracts/limitOrder/LimitOrderBook.sol [hunk 6]
```diff
@@ -219,4 +228,66 @@ contract LimitOrderBook is
...
+            if (order.isBaseToQuote) {
+                // LOB_STLOTPNM: Sell Take-profit Limit Order Trigger Price Not Matched
+                require(triggeredPrice >= order.triggerPrice, "LOB_STLOTPNM");
+            } else {
+                // LOB_BTLOTPNM: Buy Take-profit Limit Order Trigger Price Not Matched
+                require(triggeredPrice <= order.triggerPrice, "LOB_BTLOTPNM");
+            }
+        } else {
+            // LOB_IOT: Invalid Order Type
+            revert("LOB_IOT");
```
Comment (tienshaoku) [inline contracts/limitOrder/LimitOrderBook.sol:290]: Wait doesn't this violate the condition on `_verifyTriggerPrice`? https://github.com/perpetual-protocol/perp-curie-periphery-contract/pull/14/files#r887573873

Comment (tienshaoku) [inline contracts/limitOrder/LimitOrderBook.sol:290]: Perhaps instead of reverting here and adding the condition check at line `124`, what about checking `order.orderType != ILimitOrderBook.OrderType.LimitOrder` inside `_ verifySigner` and return if false?
We can thus save checking for the same thing at two places

File: contracts/limitOrder/LimitOrderBook.sol (hunk 2 above)
Comment (vinta) [reply contracts/limitOrder/LimitOrderBook.sol:124]: No, no concern. Though I prefer explicit if conditions, but no strong opinion on this. Will fix.

File: contracts/limitOrder/LimitOrderBook.sol (hunk 3 above)
Comment (vinta) [reply contracts/limitOrder/LimitOrderBook.sol:253]: Ah, backend service used to call `verifySigner()` to do an extra check when storing orders into DB, but it didn't do that now.

Can make both functions public or internal. I think public is ok, they're useful utility functions sometimes. What do you prefer?

File: contracts/interface/ILimitOrderBook.sol [hunk 7]
```diff
@@ -6,7 +6,8 @@ interface ILimitOrderBook {
     // Do NOT change the order of enum values because it will break backwards compatibility
     enum OrderType {
         LimitOrder,
-        StopLimitOrder
+        StopLimitOrder,
```
Comment (tienshaoku) [inline contracts/interface/ILimitOrderBook.sol:9]: Wonder why aren't they `stopLoss` & `takeProfit`? I thought these two are symmetric

File: contracts/limitOrder/LimitOrderBook.sol [hunk 8]
```diff
@@ -219,4 +228,66 @@ contract LimitOrderBook is
...
+        // NOTE: we can only support stop/take-profit limit order for markets that use ChainlinkPriceFeed
+        uint256 triggeredPrice = _getPriceByRoundId(order.baseToken, roundIdWhenTriggered);
+
+        // we need to make sure the price has reached trigger price.
+        // however, we can only know whether index price has reached trigger price,
+        // we didn't know whether market price has reached trigger price
+
+        // rules of advanced order types
+        // https://help.ftx.com/hc/en-us/articles/360031896592-Advanced-Order-Types
+        if (order.orderType == ILimitOrderBook.OrderType.StopLimitOrder) {
```
Comment (tienshaoku) [inline contracts/limitOrder/LimitOrderBook.sol:272]: I guess we need only two cases? cuz bot runners can tell the order type from the order itself?

File: contracts/limitOrder/LimitOrderBook.sol [hunk 9]
```diff
@@ -219,4 +228,66 @@ contract LimitOrderBook is
...
+        // LOB_IRI: Invalid RoundId
+        require(order.roundIdWhenCreated > 0 && roundIdWhenTriggered > order.roundIdWhenCreated, "LOB_IRI");
+
+        // LOB_ITP: Invalid Trigger Price
+        require(order.triggerPrice > 0, "LOB_ITP");
+
+        // NOTE: we can only support stop/take-profit limit order for markets that use ChainlinkPriceFeed
+        uint256 triggeredPrice = _getPriceByRoundId(order.baseToken, roundIdWhenTriggered);
+
+        // we need to make sure the price has reached trigger price.
```
Comment (tienshaoku) [inline contracts/limitOrder/LimitOrderBook.sol:266]: Not sure if this issue is fully discussed and I have no idea which price is used to be compared to `triggerPrice` on other exchanges

File: contracts/limitOrder/LimitOrderBook.sol [hunk 10]
```diff
@@ -112,13 +113,21 @@ contract LimitOrderBook is
         if (order.reduceOnly) {
```
Comment (tienshaoku) [inline contracts/limitOrder/LimitOrderBook.sol:113]: What about merging this if block with the other `if (order.reduceOnly)` block in line `145`? 
It's cleaner and I think an order will only be filled after a bot `callStatic` and the tx doesn't revert!

File: contracts/limitOrder/LimitOrderBook.sol [hunk 11]
```diff
@@ -219,4 +228,66 @@ contract LimitOrderBook is
 
         return signer;
     }
+
+    //
+    // INTERNAL VIEW
+    //
+
+    function _formatDecimals(
```
Comment (tienshaoku) [inline contracts/limitOrder/LimitOrderBook.sol:236]: `pure` function usually comes the last

File: contracts/limitOrder/LimitOrderBook.sol [hunk 12]
```diff
@@ -219,4 +228,66 @@ contract LimitOrderBook is
...
+        uint8 fromDecimals,
+        uint8 toDecimals
+    ) internal pure returns (uint256) {
+        // LOB_ID: Invalid Decimals
+        require(fromDecimals <= toDecimals, "LOB_ID");
+
+        return price.mul(10**(toDecimals.sub(fromDecimals)));
+    }
+
+    function _getPriceByRoundId(address baseToken, uint80 roundId) internal view returns (uint256) {
```
Comment (tienshaoku) [inline contracts/limitOrder/LimitOrderBook.sol:247]: I usually place a function used (`_getPriceByRoundId `) in another (`_verifyTriggerPrice`) behind but it's only my personal preference!

File: contracts/limitOrder/LimitOrderBook.sol (hunk 2 above)
Comment (tienshaoku) [reply contracts/limitOrder/LimitOrderBook.sol:124]: Related: https://github.com/perpetual-protocol/perp-curie-periphery-contract/pull/14/files#r887582410

File: contracts/limitOrder/LimitOrderBook.sol (hunk 8 above)
Comment (tienshaoku) [reply contracts/limitOrder/LimitOrderBook.sol:272]: Maybe even one... hmm... I have no idea and am just asking for opinions!

File: contracts/interface/ILimitOrderBook.sol [hunk 13]
```diff
@@ -99,8 +98,8 @@ interface ILimitOrderBook {
 
     /// @param order LimitOrder struct
     /// @param signature The EIP-712 signature of `order` generated from `eth_signTypedData_V4`
-    /// @param roundIdWhenTriggered The oracle `roundId` when triggerPrice is satisfied
-    // Only avaliable if orderType is StopLimitOrder, otherwise set to 0
+    /// @param roundIdWhenTriggered Chainlink `roundId` when triggerPrice is satisfied
+    // Only avaliable if orderType is StopLimitOrder/TakeProfitLimitOrder, otherwise set to 0
     function fillLimitOrder(
         LimitOrder memory order,
         bytes memory signature,
```
Comment (tienshaoku) [inline contracts/interface/ILimitOrderBook.sol:105]: Also, curious about should we place `getOrderHash` & `verifySigner` in the interface?

File: test/limitOrderBook/limitOrderBook.fillLimitOrder.triggerPrice.ts [hunk 14]
```diff
@@ -0,0 +1,457 @@
...
+            const signature = await getSignature(fixture, stopLimitOrder, trader)
+
+            await expect(
+                limitOrderBook.connect(keeper).fillLimitOrder(stopLimitOrder, signature, computeRoundId(1, 2)),
+            ).to.revertedWith("LOB_ITP")
+        })
+
+        // TODO
+        it("force error, roundId isn't existed", async () => {})
+        it("force error, baseToken isn't using ChainlinkPriceFeed", async () => {})
```
Comment (tienshaoku) [inline test/limitOrderBook/limitOrderBook.fillLimitOrder.triggerPrice.ts:203]: This one seems pretty important and we'll have to do it; perhaps at least knowing what error message will be emitted and it won't change each time!

File: test/limitOrderBook/limitOrderBook.fillLimitOrder.triggerPrice.ts [hunk 15]
```diff
@@ -0,0 +1,457 @@
...
+        })
+    })
+
+    describe("take profit limit order", async () => {
+        let currentTime: number
+
+        beforeEach(async () => {
+            currentTime = (await waffle.provider.getBlock("latest")).timestamp
+
+            await setRoundData(mockedBaseAggregator, computeRoundId(1, 1), "2700", currentTime)
```
Comment (tienshaoku) [inline test/limitOrderBook/limitOrderBook.fillLimitOrder.triggerPrice.ts:336]: Can merge this with the above `beforeEach` block.
Apart from this, tests look nice to me!

File: contracts/interface/ILimitOrderBook.sol (hunk 7 above)
Comment (vinta) [reply contracts/interface/ILimitOrderBook.sol:9]: Haha, I simply follow the naming from FTX.

Screen Shot 2022-06-06 at 14 33 05

Sure, I will rename `StopLimitOrder` to `StopLossLimitOrder` in contract.

File: contracts/limitOrder/LimitOrderBook.sol (hunk 8 above)
Comment (vinta) [reply contracts/limitOrder/LimitOrderBook.sol:272]: Since there has to be 4 `require` statements, so what's the advantage to only have 1 or 2 error messages?

I consider this is similar to https://github.com/perpetual-protocol/perp-lushan/blob/main/contracts/ClearingHouse.sol#L1069

File: contracts/interface/ILimitOrderBook.sol (hunk 13 above)
Comment (vinta) [reply contracts/interface/ILimitOrderBook.sol:105]: It looks like the compiler only allows external functions in the interface, but `getOrderHash` is public. Is it ok to write `function getOrderHash(LimitOrder memory order) external view returns (bytes32)`?

Compiler passes though, so I added it anyway.

File: test/limitOrderBook/limitOrderBook.fillLimitOrder.triggerPrice.ts (hunk 14 above)
Comment (vinta) [reply test/limitOrderBook/limitOrderBook.fillLimitOrder.triggerPrice.ts]: Yes, added! Please review https://github.com/perpetual-protocol/perp-curie-periphery-contract/pull/14/commits/e6dd162c2be00b83cfcdd8cdf6b07f1b38350f1a

File: contracts/limitOrder/LimitOrderBook.sol (hunk 3 above)
Comment (vinta) [reply contracts/limitOrder/LimitOrderBook.sol:236]: Decided to make both `verifySigner()` and `verifyTriggerPrice()` internal (so I don't have to write tests for them).

File: contracts/limitOrder/LimitOrderBook.sol (hunk 5 above)
Comment (vinta) [reply contracts/limitOrder/LimitOrderBook.sol:40]: Yes, we could. Though I think this is less error-prone since we might forget to change the hash somehow. Also more readable than `54ea8c184890b5a7f7321f45a2ae952f0af50e3467b2216418a683566bf57c30`.

File: contracts/limitOrder/LimitOrderBook.sol (hunk 9 above)
Comment (vinta) [reply contracts/limitOrder/LimitOrderBook.sol:249]: There will be 2 phases to fill a stop-loss/take-profit limit order:

Phase 1:

The [keeper](https://github.com/perpetual-protocol/perp-curie-limit-order/tree/main/packages/keeper) will keep tracking index price changes (through Chainlink aggregator events), and compare the index price with limit orders' trigger price. If index price matches trigger price (the logic should be similar to [this](https://github.com/perpetual-protocol/perp-curie-periphery-contract/blob/feature/stop-limit-order/contracts/limitOrder/LimitOrderBook.sol#L255) I suppose), the keeper will save the index price's `roundId` into `roundIdWhenTriggered` field in DB.

Phase 2:

The keeper fetches limit orders with `roundIdWhenTriggered != 0`, try calling `LimitOrderBook.fillLimitOrder(order, roundIdWhenTriggered)` from time to time.

File: contracts/limitOrder/LimitOrderBook.sol (hunk 1 above)
Comment (vinta) [reply contracts/limitOrder/LimitOrderBook.sol:247]: Done!

File: contracts/limitOrder/LimitOrderBook.sol (hunk 5 above)
Comment (tienshaoku) [reply contracts/limitOrder/LimitOrderBook.sol:40]: Hmm but I think the point of pre-computing hash is to save the gas of calculating that line `keccak256(..)`.
But yeah you're right that this might be prone to error...

I guess we can keep a line of comment as 
```
keccak256("LimitOrder(uint8 orderType,uint256 salt,address trader,address baseToken,bool isBaseToQuote,bool isExactInput,uint256 amount,uint256 oppositeAmountBound,uint256 deadline,uint160 sqrtPriceLimitX96,bytes32 referralCode,bool reduceOnly,uint80 roundIdWhenCreated,uint256 triggerPrice)")
```
and assign the value as `54ea8c184890b5a7f7321f45a2ae952f0af50e3467b2216418a683566bf57c30` directly **AND open a ticket for the deployer to check them being the same before deploying**!

Comment (vinta) [reply contracts/limitOrder/LimitOrderBook.sol:40]: You mean to save the gas when deploying or when filling orders? Or both?

File: contracts/limitOrder/LimitOrderBook.sol (hunk 8 above)
Comment (tienshaoku) [reply contracts/limitOrder/LimitOrderBook.sol:255]: Good question! 
I was thinking that since `fillLimitOrder` is most likely to be called by bots only, we won't be showing these error codes on FE and thus we don't need these 4 variances to let users know what's the exact mistake they make (on FE).

Hence, if the main audience of these error codes is code-writers, perhaps one error code telling something like `trigger price not satisfied` is sufficient... but I don't know; if you're the bot writer, will you need four of them or is one sufficient?

File: contracts/limitOrder/LimitOrderBook.sol (hunk 5 above)
Comment (tienshaoku) [reply contracts/limitOrder/LimitOrderBook.sol:40]: Both!

File: test/limitOrderBook/limitOrderBook.fillLimitOrder.triggerPrice.ts (hunk 14 above)
Comment (tienshaoku) [reply test/limitOrderBook/limitOrderBook.fillLimitOrder.triggerPrice.ts]: No I haven't haha doing it now

Comment (tienshaoku) [reply test/limitOrderBook/limitOrderBook.fillLimitOrder.triggerPrice.ts]: Done; great job!

File: contracts/limitOrder/LimitOrderBook.sol (hunk 5 above)
Comment (tienshaoku) [reply contracts/limitOrder/LimitOrderBook.sol:40]:
```
bytes32 public constant LIMIT_ORDER_TYPEHASH = 0x54ea8c184890b5a7f7321f45a2ae952f0af50e3467b2216418a683566bf57c30;
```

1. No double quote
2. Prefix `0x`

**Review by tienshaoku — APPROVED** [review]

File: contracts/limitOrder/LimitOrderBook.sol (hunk 8 above)
Comment (tienshaoku) [reply contracts/limitOrder/LimitOrderBook.sol]: Conclusion: better precision!

File: contracts/limitOrder/LimitOrderBook.sol [hunk 16]
```diff
@@ -219,4 +228,63 @@ contract LimitOrderBook is
...
+    function _verifyTriggerPrice(LimitOrder memory order, uint80 roundIdWhenTriggered) internal view {
+        if (order.orderType == ILimitOrderBook.OrderType.LimitOrder) {
+            return;
+        }
+
+        // NOTE: Chainlink proxy's roundId is always increased
+        // https://docs.chain.link/docs/historical-price-data/
+
+        // LOB_IRI: Invalid RoundId
+        require(order.roundIdWhenCreated > 0 && roundIdWhenTriggered > order.roundIdWhenCreated, "LOB_IRI");
```
Comment (paco0x) [inline contracts/limitOrder/LimitOrderBook.sol:241]: Is it possible that `roundIdWhenTriggered == order.roundIdWhenCreated`?

**Review by paco0x — APPROVED** [review]

File: contracts/limitOrder/LimitOrderBook.sol (hunk 16 above)
Comment (vinta) [reply contracts/limitOrder/LimitOrderBook.sol:241]: Ah yes, I think it's possible. Will change to `roundIdWhenTriggered >= order.roundIdWhenCreated`.
